### PR TITLE
UPSTREAM: <carry>: openshift: return no nodegroup when scaling size is zero

### PR DESCRIPTION
**Pull request by frobware**
Change nodeGroupForNode() to return nil when the underlying node group
has no scaling bounds. The underlying MachineSet in these cases is
either missing the following annotations:

```
annotations:
  machine.openshift.io/cluster-api-autoscaler-node-group-min-size: "1"
  machine.openshift.io/cluster-api-autoscaler-node-group-max-size: "6"
```

or they are present but the result of `max-size - min-size` is zero.

This change helps to remove a lot of spurious error messages from the
core of the autoscaler logs at runtime. Without this change we see the
following printed every time the autoscaler scans the state of the
cluster (which defaults to 10s) and can be very misleading.

```console
$ oc logs -f cluster-autoscaler-default-77f884d74b-527bq
E1204 08:26:01.996252       1 utils.go:467] Error while checking node group size worker-us-east-2a: group size not found
E1204 08:26:01.996545       1 utils.go:467] Error while checking node group size worker-us-east-2b: group size not found
E1204 08:26:01.996678       1 utils.go:467] Error while checking node group size worker-us-east-2c: group size not found
```

Those unit tests that were previously relying on nodeGroupForNode() to
always return a node group when no bounds had been specified have been
modified so that they now have a scaling bound >= 1.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1656334

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller.go [hunk 1]
```diff
@@ -415,6 +421,12 @@ func (c *machineController) nodeGroupForNode(node *apiv1.Node) (cloudprovider.No
 		return nil, fmt.Errorf("failed to build nodegroup for node %q: %v", node.Name, err)
 	}
 
+	// We don't scale from 0 so nodes must belong to a nodegroup
+	// that has a scale size of at least 1.
+	if nodegroup.MaxSize()-nodegroup.MinSize() < 1 {
```
Comment (ingvagabund) [inline cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller.go]: Can you also add a testcase where you set both sizes to be equal and verify you get nil node group?

Comment (frobware) [reply cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller.go]: Done in https://github.com/openshift/kubernetes-autoscaler/pull/42/commits/eec7f96068779625b486818609d6e3b0f761f256

Comment (frobware) [reply cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller.go]: And again in https://github.com/openshift/kubernetes-autoscaler/pull/42/commits/c3dcd129b96f380dafaf05256ac34e16356adfce

Comment (frobware) [reply cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller.go]: Test cases now covered in `TestControllerNodeGroupForNodeLookup()` - https://github.com/openshift/kubernetes-autoscaler/pull/42/commits/56bc385435166024ea03dd2fd020327680e96229

**Comment by ingvagabund** [comment]
/lgtm

Thanks for the improvements

**Pull request merged**